File: app/src/main/java/com/mx/gillustrated/activity/EventsActivity.java
package com.mx.gillustrated.activity;

import java.util.ArrayList;
import java.util.List;

import com.mx.gillustrated.R;
import com.mx.gillustrated.adapter.EventsAdapter;
import com.mx.gillustrated.listener.ListenerListViewScrollHandler;
import com.mx.gillustrated.vo.EventInfo;
import android.content.Intent;
import android.os.Bundle;
import android.os.Handler;
import android.os.Message;
import android.view.View;
import android.view.Window;
import android.widget.ListView;
import android.widget.RelativeLayout;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import butterknife.OnItemClick;

public class EventsActivity extends BaseActivity {

    @OnClick(R.id.btnDespairAdd)
    void onAddBtnClickListerner(){
        Intent intent = new Intent(EventsActivity.this, EventInfoActivity.class);
        intent.putExtra("event", 0);
        intent.putExtra("game", mGameId);
        startActivity(intent);
    }

    @BindView(R.id.lvDespairMain) ListView mLvDespairMain;
    @BindView(R.id.pageVBox) RelativeLayout pageVboxLayout;
    @OnItemClick(R.id.lvDespairMain)
    void listItemClickHandler(int position){
        Intent intent = new Intent(EventsActivity.this, EventInfoActivity.class);
        intent.putExtra("event", mList.get(position).getId());
        intent.putExtra("game", mGameId);
        startActivity(intent);
    }

    private EventsAdapter mAdapter;
    private List<EventInfo> mList;
    private int mGameId;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        requestWindowFeature(Window.FEATURE_NO_TITLE);
        setContentView(R.layout.activity_events);
        ButterKnife.bind(this);

        mGameId = getIntent().getIntExtra("game", 0);
        pageVboxLayout.setVisibility(View.GONE);
        mLvDespairMain.setOnScrollListener(new ListenerListViewScrollHandler(mLvDespairMain, pageVboxLayout));
        mList = new ArrayList<EventInfo>();
        mAdapter = new EventsAdapter(this, mList);

        searchMain();

    }
//
//    DespairTouchListener despairTouchListener = new DespairTouchListener(){
//
//        @Override
//        public void onSaveBtnClickListener(EventInfo info) {
//            EventInfo request = new EventInfo();
//            request.setId(info.getId());
//            request.setName(info.getName());
//            request.setGameId(mGameId);
//            long result = mDBHelper.updateEvent(request);
//            if( result > -1) {
//                Toast.makeText(EventsActivity.this, "更新成功", Toast.LENGTH_SHORT).show();
//                searchMain();
//            }
//
//        }
//    };


    Handler mainHandler = new Handler() {
        @Override
        public void handleMessage(Message msg) {
            if(msg.what == 1){
                mList.clear();
                mList.addAll((List<EventInfo>) msg.obj);
                updateList(true);
            }
        }

    };

    private void searchMain(){
        mainHandler.post( new Runnable() {
            @Override
            public void run() {
                List<EventInfo> list = mOrmHelper.getEventInfoDao().getListByGameId(mGameId, null);
                Message msg = mainHandler.obtainMessage();
                msg.what = 1;
                msg.obj = list;
                mainHandler.sendMessage(msg);
            }
        });
    }

    private void updateList(boolean flag) {
        if (flag)
            mLvDespairMain.setAdapter(mAdapter);
        else
            mAdapter.notifyDataSetChanged();
    }

}

File: app/src/main/java/com/mx/gillustrated/activity/EventsActivity.kt
package com.mx.gillustrated.activity

import com.mx.gillustrated.adapter.EventsAdapter
import com.mx.gillustrated.listener.ListenerListViewScrollHandler
import com.mx.gillustrated.vo.EventInfo
import android.content.Intent
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.os.Message
import android.view.View
import android.view.Window
import com.mx.gillustrated.databinding.ActivityEventsBinding
import java.lang.ref.WeakReference

class EventsActivity : BaseActivity() {



    private var mAdapter: EventsAdapter? = null
    private var mList: MutableList<EventInfo> = mutableListOf()
    private var mGameId: Int = 0
    private val mainHandler = ManiHandler(this)
    lateinit var binding:ActivityEventsBinding



    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        requestWindowFeature(Window.FEATURE_NO_TITLE)
        binding = ActivityEventsBinding.inflate(layoutInflater)
        setContentView(binding.root)

        mGameId = intent.getIntExtra("game", 0)
        binding.pageVBox.visibility = View.GONE
        binding.lvDespairMain.setOnScrollListener(ListenerListViewScrollHandler(binding.lvDespairMain, binding.pageVBox))
        mAdapter = EventsAdapter(this, mList)
        binding.btnDespairAdd.setOnClickListener {
            val intent = Intent(this@EventsActivity, EventInfoActivity::class.java)
            intent.putExtra("event", 0)
            intent.putExtra("game", mGameId)
            startActivity(intent)
        }
        binding.lvDespairMain.setOnItemClickListener { _, _, position, _ ->
            val intent = Intent(this@EventsActivity, EventInfoActivity::class.java)
            intent.putExtra("event", mList[position].id)
            intent.putExtra("game", mGameId)
            startActivity(intent)
        }
        searchMain()
    }

    private fun searchMain() {
        mainHandler.post {
            val list = mOrmHelper.eventInfoDao.getListByGameId(mGameId, null)
            val msg = mainHandler.obtainMessage()
            msg.what = 1
            msg.obj = list
            mainHandler.sendMessage(msg)
        }
    }

    private fun updateList(flag: Boolean) {
        if (flag)
            binding.lvDespairMain.adapter = mAdapter
        else {
            mAdapter!!.notifyDataSetChanged()
        }
    }

    companion object {

        @Suppress("UNCHECKED_CAST")
        internal class ManiHandler(eventsActivity: EventsActivity) : Handler(Looper.getMainLooper()){

            private val weakReference: WeakReference<EventsActivity> = WeakReference(eventsActivity)

            override fun handleMessage(msg: Message) {
                val activity = weakReference.get()!!
                if (msg.what == 1) {
                    activity.mList.clear()
                    activity.mList.addAll(msg.obj as List<EventInfo>)
                    activity.updateList(true)
                }
            }
        }
    }

}

